proposals/0004-FaultTolerance.md

```diff
@@ -0,0 +1,111 @@
+
+## Motivation
+
+Currently there are at least two libraries to provide fault tolerance. It is best to uniform the technologies and define a standard so that micro service applications can adopt and the implementation of fault tolerance can be provided by the containers if possible.
+
+## Proposed solution
+
+Separate the responsibility of executing logic (Runnables/Callables/etc) from guiding when execution should take place (through retry policies, bulkheads, circuit breakers). In this way, failure handling strategies become configuration that can influence executions, and the execution API itself is just responsible for receiving some configuration and performing executions.
+
+By default, a failure handling strategy could assume, for example, that any exception is a failure. But in some cases, a `null` or negative return value could also be considered a failure. Users should be able to define this, and a user's definition of a failure is what should influence execution. (This all is what the Failsafe RetryPolicy's `retryOn`, `retryWhen`, `abortIf`, etc methods are all about - defining a failure).
+
+Standardise the Fallback, Bulkhead and CircuitBreaker APIs and provide implementations.
+
+* A Java API to privide RetryPolicy
+* A Java API to provide FallBack
+* A Java API to provide BulkHead
+* A Java API to provide CircuitBreaker
+
+## Detailed design
+
+The Design is built on CDI programming model. An instance of `Execution`, `RetryPolicy`, `CircuitBreaker` and `BulkHeader` can be injected to the clients and then they can be configured as follows.
+
+### RetryPolicy: A policy define the retry criteria
+
+
+```
+RetryPolicy rp = retryPolicy.retryOn(TimeOutException.class)
+  .withDelay(2, TimeUnit.SECONDS)
+  .withMaxRetries(2);
+```
+
+When `TimeOutException` was received, delay 2 seconds and then retry 2 more times.
+
+```
+Connection connection = execution.with(rp).withFallBack(null).get(this::connect)
+```
+
+The above suppress the `TimeOutException` and provide a default result.
+
+### CircuitBreaker: a rule to define when to close the circuit
+
+```
+CircuitBreaker cb = circuitBreaker
+  .withFailureThreshold(3, 10)
+  .withSuccessThreshold(5)
+  .withDelay(1, TimeUnit.MINUTES);
+Connection connect = execution.with(cb).run(this::connect);  
+```
+
+When 3 of 10 execution failures occurs on a circuit breaker, the circuit is opened and further execution requests fail with `CircuitBreakerOpenException`. After a delay of 1 min, the circuit is half-opened and trail executions are attempted to determine whether the circuit should be closed or opened again. If the trial executions exceed a success threshold 5, the breaker is closed again and executions will proceed as normal.
+
+### BulkHead
+
+```
+BulkHead bh = bulkHead.withPool("myPool");
+Connection connect = execution.with(bh).run(this::connect);
+```
+
+Bulkhead provides a thread pool with a fixed number of threads in order to achieve thread and failure isolation.
```

Comment:
@viktorklang Assuming you're talking about attaching fault tolerance config to the operation, it could be, but overall it just depends on what the execution is and what the config is. For example, an execution (which prior to the terminal method, is just some combination of config) could be reused for Runnables that require the same kind of fault tolerance strategy:

```
Execution execution = new Execution().with(retryPolicy);
// ...
execution.run(this::openChannel);
execution.run(this::sendFile);
execution.run(this::closeChannel);
```

I'm not sure there's a lot of usefulness in that, but FWIW I've seen it done (with Failsafe).

The rationale behind the API as it is, was that it (hopefully) made for the best flow. Alternatively we could associate the config with the Runnable, but it would probably be more awkward. Maybe along the lines of this?

```
Execution.run(Executable.with(retryPolicy).and(runnable));
```

Note: in this case, Execution could be stateless.

Another alternative is we could more explicitly refer to a set of fault tolerance config as what it is, and pass it in along with the Runnable. Ex:

```
FaultToleranceConfig ftc = new FaultToleranceConfig().with(retryPolicy).with(circuitBreaker);
Execution.run(runnable, ftc);
```

Is that more of what you were thinking?